File: cpp/2359.find-closest-node-to-given-two-nodes.cpp
#include <vector>
using std::vector;

class Solution {
private:
  vector<int> dfs(const vector<int> &edges, int node) {
    vector<int> distances(edges.size(), -1);
    int dist = 0;
    distances[node] = dist;

    while (edges[node] != -1 && distances[edges[node]] == -1) {
      dist++;
      node = edges[node];
      distances[node] = dist;
    }

    return distances;
  }

public:
  int closestMeetingNode(vector<int> &edges, int node1, int node2) {
    vector<int> distances1 = dfs(edges, node1);
    vector<int> distances2 = dfs(edges, node2);

    int minDist = edges.size();
    int minNode = -1;

    for (int i = 0; i < distances1.size(); i++) {
      if (distances1[i] == -1 || distances2[i] == -1) {
        continue;
      }

      int dist = std::max(distances1[i], distances2[i]);
      if (dist < minDist) {
        minDist = dist;
        minNode = i;
      }
    }

    return minNode;
  }
};

File: python/2359.find-closest-node-to-given-two-nodes.py


class Solution:
    def closestMeetingNode(self, edges: list[int], node1: int, node2: int) -> int:
        def dfs(node: int, edges: list[int]) -> list[int]:
            n = len(edges)
            dist = [-1] * n
            dist[node] = 0

            d = 0

            # while there is next node, and there is no cycle (next node is not visited)
            while edges[node] != -1 and dist[edges[node]] == -1:
                d += 1
                node = edges[node]
                dist[node] = d

            return dist

        dist1 = dfs(node1, edges)
        dist2 = dfs(node2, edges)

        minDist = len(edges)
        minNode = -1

        for node in range(len(dist1)):
            if dist1[node] == -1 or dist2[node] == -1:
                continue

            if max(dist1[node], dist2[node]) < minDist:
                minDist = max(dist1[node], dist2[node])
                minNode = node

        return minNode

